### PR TITLE
[OSD-25656] Enhance the affinity rules to avoid pod mis-scheduling

### PR DESCRIPTION
**Pull request by xiaoyu74**
### What type of PR is this?

_(bug)_

### What this PR does / Why we need it?
Enhance the affinity rules to avoid metrics-forwarder pod mis-scheduling.

### Which Jira/Github issue(s) does this PR fix?
https://issues.redhat.com/browse/OSD-25656

### For reviewer
Based on the ticket description, the current logic to determine if the metrics-forwarder pod is mis-scheduled is:

#### 1. Check the node label:
- For each node running a metrics-forwarder pod, check the value of the hypershift.openshift.io/cluster label

#### 2. Compare namespace and serving node label:
- Compare the namespace of the metrics-forwarder pod to the value of the hypershift.openshift.io/cluster label on the serving node.
- If the namespace of the pod does not match the value of the hypershift.openshift.io/cluster label, the pod is considered mis-scheduled.

So, to resolve the mis-scheduling issue, I just enforce strict affinity rules in the metrics-forwarder deployment config to ensure the pod is scheduled only on nodes with the hypershift.openshift.io/cluster label matching the pod's namespace. 

Also, based on the test outcome, I can see there is case that many serving nodes have the hypershift.openshift.io/cluster label missing, we need to provide a affinity configuration that accommodates these scenarios as well.

Basically, in the PR, I just:
- Added required "requiredDuringSchedulingIgnoredDuringExecution" with the hypershift.openshift.io/cluster label.
- Keep the previous affinity rules for handling the label missing or other edge cases.

**P.S** More context and test outcome is in the Jira ticket [OSD-25656](https://issues.redhat.com//browse/OSD-25656)

### CODE REVIEW

File: package/resources.yaml.gotmpl [hunk 1]
```diff
@@ -154,6 +154,14 @@ spec:
...
       affinity:
         nodeAffinity:
+	  requiredDuringSchedulingIgnoredDuringExecution:
+          # Enforce scheduling on nodes with the correct hypershift.openshift.io/cluster label
+          nodeSelectorTerms:
+	  - matchExpressions:
+            - key: hypershift.openshift.io/cluster
+	      operator: In
+	      values:
+	      - '{{.package.metadata.namespace}}'
```
Comment (bmeng) [inline package/resources.yaml.gotmpl:164]: The indents here are incorrect

**Comment by xiaoyu74** [comment]
- Tested to create the deployment on the MC  `hs-mc-13vnfh7o0` in the namespace

```
oc get deployment -n ocm-staging-2eigjr6f5uj8u7l6ie5ql5656ohk1nbr-rosa-hcp-new | grep -i metrics
metrics-forwarder                         2/2     2            2           46d
metrics-forwarder-2                       1/2     2            1           52s
```

- The newly added affinity rule can be present properly.

```
oc get deployment metrics-forwarder-2 -n ocm-staging-2eigjr6f5uj8u7l6ie5ql5656ohk1nbr-rosa-hcp-new -o json | jq .spec.template.spec.affinity.nodeAffinity
{
  "preferredDuringSchedulingIgnoredDuringExecution": [
    {
      "preference": {
        "matchExpressions": [
          {
            "key": "hypershift.openshift.io/control-plane",
            "operator": "In",
            "values": [
              "true"
            ]
          }
        ]
      },
      "weight": 50
    },
    {
      "preference": {
        "matchExpressions": [
          {
            "key": "hypershift.openshift.io/request-serving-component",
            "operator": "In",
            "values": [
              "true"
            ]
          }
        ]
      },
      "weight": 100
    },
    {
      "preference": {
        "matchExpressions": [
          {
            "key": "node-role.kubernetes.io/worker",
            "operator": "Exists"
          }
        ]
      },
      "weight": 30
    }
  ],
  "requiredDuringSchedulingIgnoredDuringExecution": {
    "nodeSelectorTerms": [
      {
        "matchExpressions": [
          {
            "key": "hypershift.openshift.io/cluster",
            "operator": "In",
            "values": [
              "ocm-staging-2eigjr6f5uj8u7l6ie5ql5656ohk1nbr-rosa-hcp-new"
            ]
          }
        ]
      }
    ]
  }
}
```

**Comment by xiaoyu74** [comment]
I created a script to loop through all the HCP namespaces on the MC (`hs-mc-13vnfh7o0`) and identified that at least one of the serving nodes hosting the control-plane workloads in each HCP namespace has both the labels `'hypershift.openshift.io/cluster=ocm-staging-{{.package.metadata.namespace}}' and 'hypershift.openshift.io/request-serving-component=true'` present in pairs.

Based on this, the updated logic sets the `nodeAffinity` for scheduling `metrics-forward` pods on nodes that have both required labels. Additionally, the `preferredDuringSchedulingIgnoredDuringExecution` section has been updated with different weights to improve robustness in cases where the required labels are missing. cc: @bmeng

**Comment by xiaoyu74** [comment]
Tested with the latest updated deployment in `ocm-staging-2ghpvdic53g31liagccsfku4vphupjoi-bmeng-share` on MC `hs-mc-13vnfh7o0`

- metrics-forwarder-4.yaml
```
 cat metrics-forwarder-4.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: metrics-forwarder-4
  labels:
    app: metrics-forwarder-4
  annotations:
    package-operator.run/phase: deploy
spec:
  selector:
    matchLabels:
      app: metrics-forwarder-4
  replicas: 2
  template:
    metadata:
      labels:
        app: metrics-forwarder-4
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                  - key: hypershift.openshift.io/cluster
                    operator: In
                    values:
                      - ocm-staging-2ghpvdic53g31liagccsfku4vphupjoi-bmeng-share
          preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 100
              preference:
                matchExpressions:
                  - key: hypershift.openshift.io/cluster
                    operator: In
                    values:
                      - ocm-staging-2ghpvdic53g31liagccsfku4vphupjoi-bmeng-share
                  - key: hypershift.openshift.io/request-serving-component
                    operator: In
                    values:
                      - 'true'
            - weight: 90
              preference:
                matchExpressions:
                  - key: hypershift.openshift.io/control-plane
                    operator: In
                    values:
                      - 'true'
            - weight: 80
              preference:
                matchExpressions:
                  - key: node-role.kubernetes.io/worker
                    operator: Exists
        podAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - podAffinityTerm:
                labelSelector:
                  matchLabels:
                    hypershift.openshift.io/hosted-control-plane: ocm-staging-2ghpvdic53g31liagccsfku4vphupjoi-bmeng-share
                topologyKey: kubernetes.io/hostname
              weight: 100
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchLabels:
                  app: metrics-forwarder-4
              topologyKey: topology.kubernetes.io/zone
      automountServiceAccountToken: false
      containers:
        - name: nginx
          image: registry.access.redhat.com/ubi8/nginx-120
          command:
            - nginx
            - '-g'
            - daemon off;
          ports:
            - containerPort: 8001
          volumeMounts:
            - name: nginx-config
              mountPath: /etc/nginx/nginx.conf
              subPath: nginx.conf
            - mountPath: /etc/nginx/cert
              name: nginx-cert
      tolerations:
        - effect: NoSchedule
          key: hypershift.openshift.io/control-plane
          operator: Equal
          value: 'true'
        - effect: NoSchedule
          key: hypershift.openshift.io/cluster
          operator: Equal
          value: ocm-staging-2ghpvdic53g31liagccsfku4vphupjoi-bmeng-share
        - effect: NoSchedule
          key: hypershift.openshift.io/request-serving-component
          operator: Equal
          value: 'true'
      volumes:
        - name: nginx-config
          configMap:
            defaultMode: 420
            name: metrics-forwarder-config
        - name: nginx-cert
          secret:
            defaultMode: 420
            secretName: metrics-forwarder-secret
```

- Applying the deployment to the HCP cluster
```
bp-elevate apply -f metrics-forwarder-4.yaml

oc get deploy | grep metrics
metrics-forwarder-4                       2/2     2            2           17s
```
